driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/vault/ClientEncryptionImpl.java

```diff
@@ -183,6 +198,85 @@ public Publisher<RewrapManyDataKeyResult> rewrapManyDataKey(final Bson filter, f
                         }));
     }
 
+    @Override
+    public Publisher<BsonDocument> createEncryptedCollection(final MongoDatabase database, final String collectionName,
+            final CreateCollectionOptions createCollectionOptions, final CreateEncryptedCollectionParams createEncryptedCollectionParams) {
+        notNull("collectionName", collectionName);
+        notNull("createCollectionOptions", createCollectionOptions);
+        notNull("createEncryptedCollectionParams", createEncryptedCollectionParams);
+        // this publisher captures and mutates state, which is why it has to be one-shot
+        return OneShotPublisher.from(Mono.defer(() -> {
+            MongoNamespace namespace = new MongoNamespace(database.getName(), collectionName);
+            Bson rawEncryptedFields = createCollectionOptions.getEncryptedFields();
+            if (rawEncryptedFields == null) {
+                throw new MongoConfigurationException(format("`encryptedFields` is not configured for the collection %s.", namespace));
```

Comment:
Instead of throwing here, it is likely more idiomatic to return `Mono.error`, given that we are in the `Supplier` that is passed to `Mono.defer`. However, if we to follow this approach consistently, then the whole `Supplier` passed to `Mono.defer` should be wrapped in a `try` statement that catches all non-fatal exceptions and returns `Mono.error` instead of throwing them.

But I see no reason to do that:

- it is one more thing we should always remember to do;
- it arguably makes the code less readable;
- why doing that when the Project Reactor already [wraps](https://github.com/reactor/reactor-core/blob/v3.5.0/reactor-core/src/main/java/reactor/core/publisher/MonoDefer.java#L43-L50) the `Supplier` passed to `Mono.defer` in the `try` statement and signals `onError` for us?

Reply:
I would do this work outside of the construction of the publisher. That way it follows the general validation eg `notNull` checks and reduces the code in the callable.

Reply:
I initially had it like that, but then noticed that the other `MongoConfigurationException` we throw if `fields` is not an array, should be signalled via the publisher: we need to encode the `encryptedFields` config to `BsonDocument`, and this does not seem anymore like just an argument check.

Reporting `MongoConfigurationException` from the same method in two different ways (`throw` or signal `onError)` seems weird, so I decided to also signal the first `MongoConfigurationException` that you commented on. Given this consideration, does the code seem better now?

Reply:
Based on your experimental code [here](https://github.com/mongodb/mongo-java-driver/pull/1079#issuecomment-1421137054), I infer that you still think that failures to render `Bson` should be treated as fatal and not signalled via the returned publisher. So I do this work outside of the construction of the publisher as you originally requested.